### PR TITLE
[lldb] Update dwim-print to support limited variable expression paths

### PR DESCRIPTION
**Pull request by kastiglione**
`frame variable` supports nested variable access, which the API calls "variable expression paths". This change updates `dwim-print` to support a subset of supported variable expression paths.

Consider the expression `a->b`. In C++, the arrow operator can be overloaded, and where that is the case, expression evaluation must be used to evaluate it, not frame variable. Likewise, the subscript operator can be overloaded.

To avoid those cases, this change introduces a limited support for variable expression paths. Use of the dot operator is allowed.

Additionally, this change allows `dwim-print` to directly access children of `this` and `self` (see AllowDirectIVarAccess). This functionality is also provided by the same `GetValueForVariableExpressionPath` method.

rdar://104348908

### CODE REVIEW

File: lldb/source/Commands/CommandObjectDWIMPrint.cpp [hunk 1]
```diff
@@ -151,10 +151,23 @@ void CommandObjectDWIMPrint::DoExecute(StringRef command,
...
-    if (valobj_sp && valobj_sp->GetError().Success()) {
+  // First, try `expr` as a _limited_ frame variable expression path: only the
+  // dot operator (`.`) is permitted for this case.
+  //
+  // This is limited to support only unambiguous expression paths. Of note,
+  // expression paths are not attempted if the expression contain either the
+  // arrow operator (`->`) or the subscript operator (`[]`). This is because
+  // both operators can be overloaded in C++, and could result in ambiguity in
+  // how the expression is handled. Additionally, `*` and `&` are not supported.
+  bool try_variable_path = expr.find_first_of("*&->[]") == StringRef::npos;
```
Comment (Michael137) [inline lldb/source/Commands/CommandObjectDWIMPrint.cpp:162]: Nit:
```suggestion
  const bool try_variable_path = expr.find_first_of("*&->[]") == StringRef::npos;
```

**Review by Michael137 — APPROVED** [review]
Makes sense to me!

File: lldb/test/API/commands/dwim-print/TestDWIMPrint.py [hunk 2]
```diff
@@ -154,5 +163,25 @@ def test_preserves_persistent_variables(self):
...
+
+    def test_variable_expression_path(self):
+        """Test dwim-print supports certain variable expression paths."""
+        self.build()
+        lldbutil.run_to_source_breakpoint(
+            self, "break here", lldb.SBFileSpec("main.cpp")
+        )
+        self.runCmd("settings set auto-one-line-summaries false")
+        self._expect_cmd("dwim-print w.s", "frame variable")
+        self._expect_cmd("dwim-print wp->s", "expression")
```
Comment (Michael137) [inline lldb/test/API/commands/dwim-print/TestDWIMPrint.py:179]: Should we also add cases for `*`/`[]`/`&`? Or are those tested elsewhere?

Comment (kastiglione) [reply lldb/test/API/commands/dwim-print/TestDWIMPrint.py:179]: there are tests for those in `test_variable_paths`

**Review by augusto2112 — APPROVED** [review]

File: lldb/source/Commands/CommandObjectDWIMPrint.cpp (hunk 1 above)
Comment (kastiglione) [reply lldb/source/Commands/CommandObjectDWIMPrint.cpp]: done

**Pull request merged**